File: tic-tac-toe-javafx-java/src/main/java/com/mj/tic/tac/toe/javafx/java/util/Difficulty.java
package com.mj.tic.tac.toe.javafx.java.util;

/**
 * Represents the three AI difficulty levels available in the game.
 * <p>
 * Each constant is associated with a resource-bundle key that resolves to a
 * localized display label (e.g. {@code "Easy"}, {@code "Moyen"}, etc.),
 * enabling the UI to render the difficulty name in the user's locale.
 * </p>
 *
 * <p>The difficulty level controls the computer's move-selection strategy:</p>
 * <ul>
 *   <li>{@link #EASY} — picks a random empty cell.</li>
 *   <li>{@link #MEDIUM} — randomly chooses between a random move and the
 *       optimal minimax move (50/50).</li>
 *   <li>{@link #HARD} — always plays the optimal move via the full minimax
 *       algorithm.</li>
 * </ul>
 *
 * @author Montaser Sbaih
 * @version 1.0
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @since 20-01-2023
 */
public enum Difficulty {

    EASY("control.label.difficulty.easy"),
    MEDIUM("control.label.difficulty.medium"),
    HARD("control.label.difficulty.hard");

    /**
     * Resource-bundle key for the localized display name of this difficulty.
     */
    private final String resourceKey;

    /**
     * Constructs a difficulty level constant.
     *
     * @param resourceKey the resource-bundle key used to look up the localized label.
     */
    Difficulty(String resourceKey) {
        this.resourceKey = resourceKey;
    }

    /**
     * Returns the resource-bundle key used to look up this difficulty's localized label.
     *
     * @return the resource-bundle key (e.g. {@code "control.label.difficulty.easy"}).
     */
    public String getResourceKey() {
        return resourceKey;
    }

    /**
     * Returns the resource-bundle key for this difficulty's localized display name.
     *
     * @return the resource-bundle key (e.g. {@code "control.label.difficulty.easy"}).
     */
    @Override
    public String toString() {
        return this.resourceKey;
    }
}

File: tic-tac-toe-javafx-kotlin/src/main/kotlin/com/mj/tic/tac/toe/javafx/kotlin/util/Difficulty.kt
package com.mj.tic.tac.toe.javafx.kotlin.util

/**
 * Represents the AI difficulty level selected by the player.
 *
 * Each constant determines which [MoveStrategy] implementation the
 * AI uses to compute its moves:
 * - [EASY]: always uses [RandomMoveStrategy] (purely random moves).
 * - [MEDIUM]: randomly alternates between [RandomMoveStrategy] and
 *   [MinimaxMoveStrategy] with equal probability (50/50).
 * - [HARD]: always uses [MinimaxMoveStrategy] for optimal play.
 *
 * @property resourceKey The i18n resource bundle key used to look up
 *   the localized display name for this difficulty level.
 * @author Montaser Sbaih
 * @version 1.0
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @since 29-05-2026
 */

enum class Difficulty(val resourceKey: String) {
    /** AI picks random moves using [RandomMoveStrategy]. */
    EASY("control.label.difficulty.easy"),

    /** AI randomly alternates between random and minimax strategies. */
    MEDIUM("control.label.difficulty.medium"),

    /** AI uses full minimax with alpha-beta pruning via [MinimaxMoveStrategy]. */
    HARD("control.label.difficulty.hard");

    /**
     * Returns the i18n resource key for this difficulty level.
     *
     * @return The [resourceKey] string used for localized label lookups.
     */
    override fun toString(): String = resourceKey
}
